pkg/provisioner/controller.go

```diff
@@ -291,6 +293,103 @@ func (c *obcController) handleProvisionClaim(key string, obc *v1alpha1.ObjectBuc
 	// to control access to static buckets via RBAC rules on storage classes.
 	isDynamicProvisioning := isNewBucketByStorageClass(class)
 
+	// Re-Get the claim in order to shorten the race condition where the claim was deleted after provisioning started
+	obc, err = claimForKey(key, c.libClientset)
+	if err != nil {
+		if errors.IsNotFound(err) {
+			return fmt.Errorf("OBC was lost before we could provision: %v", err)
+		}
+		return err
+	}
+
+	// Older versions of this library set obc.Spec.BucketName with the name of the bucket generated
+	// from obc.Spec.GenerateBucketName, which causes an error condition when updating to a newer
+	// version of this library if provisioning runs again. If the bucket name has the
+	// GenerateBucketName prefix, we can assume that this was originally intended to be a generated
+	// name.
+	if obc.Spec.GenerateBucketName != "" && obc.Spec.BucketName != "" {
+		if strings.HasPrefix(obc.Spec.BucketName, obc.Spec.GenerateBucketName+"-") {
+			obc.Spec.BucketName = ""
+			obc, err = updateClaim(
+				c.libClientset,
+				obc,
+				defaultRetryBaseInterval,
+				defaultRetryTimeout)
+			if err != nil {
+				return fmt.Errorf("error updating OBC to remove set spec.bucketName matching the prefix from spec.generateBucketName: %v", err)
+			}
+		}
+	}
+
+	updateOnSuccess := func() error {
+		// In the case where the OB has been created and the operator crashed immediately after
+		// (before the OB phase could be updated), we should update the OB phase in addition to
+		// updating the OBC reference to the OB and the OBC's phase.
+		ob, err = updateObjectBucketPhase(
+			c.libClientset,
+			ob,
+			v1alpha1.ObjectBucketStatusPhaseBound,
+			defaultRetryBaseInterval,
+			defaultRetryTimeout)
+		if err != nil {
+			return fmt.Errorf("error updating OB %q's status to %q: %v", ob.Name, v1alpha1.ObjectBucketStatusPhaseBound, err)
+		}
+		// update OBC
+		obc.Spec.ObjectBucketName = ob.Name
+		// Do not set obc.Spec.BucketName here: (1) if user has it set, it should be as they set it
+		// (2) if generateBucketName is set, setting this will cause errors if provisioning runs again
+		obc, err = updateClaim(
+			c.libClientset,
+			obc,
+			defaultRetryBaseInterval,
+			defaultRetryTimeout)
+		if err != nil {
+			return fmt.Errorf("error updating OBC: %v", err)
+		}
+		obc, err = updateObjectBucketClaimPhase(
+			c.libClientset,
+			obc,
+			v1alpha1.ObjectBucketClaimStatusPhaseBound,
+			defaultRetryBaseInterval,
+			defaultRetryTimeout)
+		if err != nil {
+			return fmt.Errorf("error updating OBC %q's status to: %v", v1alpha1.ObjectBucketClaimStatusPhaseBound, err)
+		}
+		return nil
+	}
+
+	// recover from degraded state where provisioning is happening again on a completed obc
+	ob, err = getObjectBucketFromClaimKey(key, c.libClientset, defaultRetryBaseInterval, defaultRetryTimeout)
+	if err != nil {
+		return fmt.Errorf("error determining if object bucket for this claim exists: %v", err)
+	}
+	if ob != nil {
```

Comment:
A bit simplified:
```
	if ob != nil {
...		if ob.Spec.ClaimRef == nil || *ob.Spec.ClaimRef == emptyRef || ob.Spec.ClaimRef.UID == types.UID("") {
			logD.Info("Attempting to re-bind OB %q with OBC %s/%s", ob.Name, obc.Namespace, obc.Name)
			ob.Spec.ClaimRef, err = claimRefForKey(key, c.libClientset)
			if err != nil {
				return fmt.Errorf("error getting OB's reference to OBC: %v", err)
			}
			ob, err = updateObjectBucket(ob, c.libClientset, defaultRetryBaseInterval, defaultRetryTimeout)
			if err != nil {
				return fmt.Errorf("error setting OB's reference to OBC: %v", err)
			}
		} else if obc.UID != ob.Spec.ClaimRef.UID {
			return fmt.Errorf("OB %q that is not owned by this OBC, which is likely an artifact from a previous OBC that wasn't cleaned up. Stopping OBC provisioning", ob.Name)
		}
		err = updateOnSuccess()
		if err != nil {
			return fmt.Errorf("error re-binding OB %q with OBC %s/%s", ob.Name, obc.Namespace, obc.Name, err)
		}
		// Do not do any more provisioning; the bucket is already provisioned and merely
		// needed re-associated to the claim
		logD.Info("re-binding OB to OBC successful")
		return nil
```

Reply:
The main point of this comment above, besides a bit shorter messages, if that I don't see the need for the `} else if obc.UID == ob.Spec.ClaimRef.UID {`

Reply:
I'd like to keep the middle `else if` line in order to have a definite log message show when the library is taking the code path where it is updating an OBC to re-bind to an existing OB for the upgrade case.

Reply:
I didn't take your suggestion word-for-word where I thought the info might be able to be refined a little. I did update the messages and made sure the OB and OBC info would get reported. LMK if you still have changes.

Reply:
Looking at the latest code, you've decided to forgo the log msgs related to re-binding and claimRef/UID checks. Is this correct of has this previous block been moved?